### PR TITLE
Activities Health Dashboard Front End

### PR DESCRIPTION
**Pull request by cissyxyu**
## WHAT
All front end code for an Activities Health Dashboard that includes:
- sub component Prompt Health Dashboard that shows individual prompt health
- all relevant data fetched from the activities health endpoint in the LMS
- filters and sorting functionality
- Download as CSV functionality
- new reducer code to include a new action "Save Activities Health Flag" to persist the flag across user actions 

## WHY
Part of an expansion of the Curriculum team's tools.

## HOW
Adding new React components to render the ActivityHealth table and it's subcomponent PromptHealth table. Adding a bunch of utility functions to help render the table.

### Screenshots
Screen Shot 2021-05-07 at 11 40 21 AM
Screen Shot 2021-05-07 at 11 40 35 AM

### Notion Card Links
https://www.notion.so/quill/Project-Card-Activity-Health-Dashboard-886c7aa8e72c45f6b4c84ad69fd30629

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  YES
Have you deployed to Staging? | YES
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Design Review: If applicable, have you compared the coded design to the mockups? | Yes

### CODE REVIEW

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+
+interface ActivityHealthState {
+  loadingTableData: boolean;
+  flag: string;
+  fetchedData: Array<Object>;
+  promptSearchInput: string;
+  dataToDownload: Array<Object>;
+}
+
+class ActivityHealth extends React.Component<ActivityHealthProps, ActivityHealthState> {
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx:48]: NBD for this PR, but as a note for going forward, we have been trying to write new React components as functional components, using the React Hook syntax instead of declaring state.

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+    dataToDownload: []
+  };
+
+  componentDidMount() {
+    this.fetchActivityHealthData();
+  }
+
+  shouldComponentUpdate(nextProps, nextState){
+    return !(this.state === nextState)
+  }
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx:64]: Why is this here?

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+        Header: 'Tool',
+        accessor: 'tool',
+        filterMethod: (filter, row) => {
+          if (filter.value === "all") {
+            return true;
+          }
+          else {
+            return row[filter.id] === filter.value
+          }
+        },
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx]: since you return from the `if` block, you can just do:

```
if (filter.value === "all") { return true }

return row[filter.id] === filter.value
```

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+        resizeable: true,
+        sortMethod: sortByList,
+        Cell: (row) => (
+          <div>
+            {
+              row.original['activity_packs'] ?
+              row.original['activity_packs'].map((ap, index) => {
+                if (!ap.name) return "";
+                else if (index != row.original['activity_packs'].length - 1) return <div key={ap.id}>{ap.name},</div>
+                else return <div key={ap.id}>{ap.name}</div>
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx]: similarly here, no need for an `else` in front of the final return.

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+    ];
+  }
+
+  addCommasToThousands = (num) =>
+  {
+    if (!num) return ""
+    var num_parts = num.toString().split(".");
+    num_parts[0] = num_parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ",");
+    return num_parts.join(".");
+  }
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx]: This doesn't need to be a component-specific method since it isn't using props or state.

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+        Cell: props => props.value,
+        maxWidth: 150
+      }
+    ];
+  }
+
+  addCommasToThousands = (num) =>
+  {
+    if (!num) return ""
+    var num_parts = num.toString().split(".");
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx]: Let's use `const` or `let` instead of var: https://github.com/airbnb/javascript#references

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+    num_parts[0] = num_parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ",");
+    return num_parts.join(".");
+  }
+
+  getToolBadge = (tool) => {
+    if (tool === CONNECT_TOOL) {
+      return (<div className="badge connect-badge">c</div>)
+    } else {
+      return (<div className="badge grammar-badge">g</div>)
+    }
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx]: no need for else

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+  }
+
+  formatTableForCSV = (e) => {
+    const currentRecords = this.reactTable.getResolvedState().sortedData
+    const columns = this.columnDefinitions()
+    let dataToDownload = []
+    for (var index = 0; index < currentRecords.length; index++) {
+       let recordToDownload = {}
+       for(var colIndex = 0; colIndex < columns.length ; colIndex ++) {
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx]: can use `let` instead of `var` in for loops

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+       this.csvLink.link.click();
+    })
+  }
+
+  handleFlagChange = (e) => {
+    const { dispatch } = this.props
+    dispatch(actions.setFlag(e.target.value))
+
+    // this is a dummy action, we dont use the state value but we need to trigger a re-render
+    this.setState({flag: e.target.value})
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx:271]: If you didn't have that `shouldComponentUpdate` method above, you wouldn't need this dummy method here because the prop change should trigger a re-render.

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/numberFilterInput.tsx [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  label: string;
+  filter: {
+    value: string;
+  }
+}
+
+interface NumberFilterInputState {
+}
+
+export default class NumberFilterInput extends React.Component<NumberFilterInputProps, NumberFilterInputState> {
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Connect/components/activityHealth/numberFilterInput.tsx]: This is begging to be a purely functional component.

File: services/QuillLMS/client/app/bundles/Connect/styles/style.scss [hunk 11]
```diff
@@ -531,3 +531,92 @@ textarea.submission {
   background-color: white;
   width: 360px;
 }
+
+.activity-health-filters {
```
Comment (emilia-friedberg) [inline services/QuillLMS/client/app/bundles/Connect/styles/style.scss]: This and the following CSS should get its own file.

**Review by emilia-friedberg — APPROVED** [review]
I made some stylistic comments, but overall looks really good. Excited to see this in production!

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx (hunk 2 above)
Comment (cissyxyu) [reply services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx:64]: I need to stop the component from re-rendering when props are changed, and only re-render when state is changed, because I want the expand/collapse menus to stay open rather than close when, for example, the user goes to another tab and then clicks back to this tab.

Not sure if this is the correct approach but it did work for my purposes. Open to other suggestions!

File: services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx (hunk 9 above)
Comment (cissyxyu) [reply services/QuillLMS/client/app/bundles/Connect/components/activityHealth/activityHealth.tsx:271]: See the comment on why I need to ignore props change above.

**Review by eadams17 — APPROVED** [review]
